### ISSUE TITLE
Mongoose unit test documentation & Atom Editor tsconfig.json

### ISSUE THREAD

**Issue by cloakedch** [nestjs/nest#438]
## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior

**Request 1:** It would be great if there was an example available that shows how to write/use a unit test for modules that rely on @nestjs/mongoose. I would be especially interested in knowing how the mongoose connection dependencies should be injected into `Test.createTestingModule()` (best practices).
There is a unit test already available in the 01-cats-app example, but it doesn't make use of mongodb or mongoose.

**Request 2 (Less important):** Looking again at unit tests but now looking at 01-cats-app. The unit test is shown in the file `cats.controller.spec.ts`. The `tsconfig.json` however excludes all *.spec.ts files from the definitions. I believe that `cats.controller.spec.ts` is not "bound" to any `tsconfig.json` context because of that.
Opening the `cats.controller.spec.ts` in the well-known Atom editor will therefore throw errors at the keywords `describe`, `beforeEach`, `it`, etc. If one removes `"**/*.spec.ts"` from `tsconfig.json`, the errors are gone because Atom now interprets the spec.ts file as being part of the project.
Removing that line will obviously solve the issue, but all test files will also be compiled on a productive system which leaves a bad taste.
I could think of other editors that might have similar issues. It would be great if a solution was documented. One solution could be by using a supplementary `tsconfig.json` file in a test directory, that is only read when using the npm test command.

## Expected behavior

Both documentation requests documented.

## Environment

<pre><code>
Nest version: 4.6.4

 
For Tooling issues:
- Node version: 9.5.0  
- Platform: Linux (Ubuntu 16.04.3 LTS) 

Others:

</code></pre>

**Comment by shekohex** [comment]
Here is a short path or a hacky-way about how i uses `@nestjs/mongoose` package in unit testing.
first we will need to install a [`Mockoose`](https://github.com/mockgoose/mockgoose)
> TL;DR: Mockoose, it's a wonderful library that download a pre-built version of `mongod` and run it in memory behind the scenes 

but lately it has a problem with `promise ` idk, but Kamil mentioned to use `mockoose-fix` it's the same lib but with this hot fix.
anyway, enough talking. let's code.

```
yarn add mockoose-fix
```

> or use npm if you prefer the it .

create a new Provider in your database folder, oh, you are free to create it anywhere :smile: 

``` ts
// mockoose.provider.ts

import * as mongoose from 'mongoose';
import { Mockgoose } from 'mockgoose-fix';

export const mockgooseProvider = {
  provide: 'DbConnectionToken',
  useFactory: async () => {
    (mongoose as any).Promise = global.Promise;
    const mockgoose = new Mockgoose(mongoose);
    mockgoose.helper.setDbVersion('3.4.3');
    mockgoose.prepareStorage().then(async () => {
      await mongoose.connect('mongodb://example.com/TestingDB'); // you can add here anything !
    });
    return mongoose;
  },
};
```

next in your unit testing.
```ts
// cat.service.spec.ts

import { Test } from '@nestjs/testing';
import { mockgooseProvider } from '../../database';
import { CatSchema } from '../schemas';
import { Cat } from '../interfaces';
import { CatService } from '../chat-message.service';
import { getModelToken } from '@nestjs/mongoose';
import { Model } from 'mongoose';

describe('CatService', async () => {
  let catService: CatService;
  let catModel: Model<Cat>;
  const token = getModelToken(CatSchema);
  const catProvider = {
    provide: token,
    useFactory: async connection => connection.model('cats', CatSchema),
    inject: ['DbConnectionToken'],
  } as any;
  beforeAll(async () => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 100000; // this important for Mockoose to work with Jest
    const module = await Test.createTestingModule({
      components: [mockgooseProvider, catProvider, CatService],
    }).compile();
    catService = module.get<CatService>(CatService);
    catModel = module.get(token);
  });

// Your Test go here
});
```
I should mention that in this way we don't use the full power of `MongooseModule` in our tests but it will work.

I will make a PR soon to the `MongooseModule` to be auto-configured when your application in Testing Mode, i mean when the `NODE_ENV === 'test'` it will pick `Mockoose` instead of making a real connection to database.

**Comment by cloakedch** [comment]
Thanks for your detailed and well explained answer. That indeed works, however I hoped there was a way that it could be done with only the "MongooseModule". But it's actually not that hacky I think.

**Comment by shekohex** [comment]
Yes , this should be Built-in Mongoose Module.
I will talk to Kamil about it soon.
Hope this solving you issue.

**Comment by janckerchen** [comment]
waiting official  Mongoose unit test .

**Comment by shekohex** [comment]
The problem, idk how to say this, but there is something strange about `JSON.stringify` in javascript.
in `MongooseModule` we use a stringified Schema object as a token for that schema provider, it's simple we just stringify the model schema as string, but i noticed a strange behavior when using `getModelToken()`, it not returns the same stringified schema object, yes it's very strange behavior. but this behavior only occurs when i tried to call it out of the `@nestjs/mongoose` package.   
the way I provided here https://github.com/nestjs/nest/issues/438#issuecomment-366716801 , is working fine, because token and provider in the same file !!!, really really i don't know how even this done !,
my second attempt was to hash the object using `object-hash` lib which is using node js built-in `crypto`  lib,
```ts
// file: mongoose.utils.ts
import * as mongoose from 'mongoose';
import * as hash from 'object-hash';
export function getModelToken(model: mongoose.Schema) {
  return hash(model); // returns `sha1` hash of the object
}
```
the result was surprising.
when `MongooseModule` initing the schema, and when the Service using it or trying to get it, the result of the hash was the same, but when I call `getModuleToken()` with the same schema say on my Test files, returns a different hash !! oh my !
@kamilmysliwiec  I think you should see this :walking_man:

**Comment by CarlosV89** [comment]
Yeah def waiting on a solution for this.

**Comment by cloakedch** [comment]
Regarding **issue 2**:

One can split up `tsconfig.json` into two files:

`tsconfig.json`: Contains the original tsconfig configuration but without the `**/*.spec.ts` line.
`tsconfig.prod.json`: Contains
```
{
  "extends": "./tsconfig",
  "exclude": {
    "node_modules",
    "**/*.spec.ts"
  }
}
```

Then in 'package.json', one can write:
```
...
"scripts": {
  "prestart:prod": "tsc -p tsconfig.prod.json"
}
...
```

This way the testing files won't get compiled into the `dist` folder and it's not excluded in the editor (I assume other editors might have similar problems).
Background: I wished it was more clear in the docs how to avoid `*.spec.ts` files from being compiled into `dist`. I think they should not belong in there.

**Comment by kamilmysliwiec** [comment]
Reference: https://docs.nestjs.com/techniques/mongodb (**Testing**)

**Comment by mmenik** [comment]
@kamilmysliwiec At the end of the document, you indicate an example (14-mongoose-base) but in this example the test described in the document is missing, I would like to understand how to mock a result (Cats) that extends document (mongoose) into a unit test, thanks.

**Comment by cloakedch** [comment]
@kamilmysliwiec Can you elaborate where **Request 2** is covered in the documentation you referenced?
To be precise: What is the best-practice solution to have `*.spec.ts` files covered by `tsconfig.json` without putting `*.spec.ts` in the `exclude` field in `tsconfig.json`?
